src/solvers/lowering/byte_operators.cpp

```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
 
 static exprt lower_byte_update(
   const byte_update_exprt &src,
-  const namespacet &ns,
-  bool negative_offset)
+  const array_exprt &value_as_byte_array,
+  const optionalt<exprt> &non_const_update_bound,
+  const namespacet &ns);
+
+static exprt lower_byte_update_byte_array_vector(
+  const byte_update_exprt &src,
+  const typet &subtype,
+  const array_exprt &value_as_byte_array,
+  const optionalt<exprt> &non_const_update_bound,
+  const namespacet &ns)
 {
-  const auto element_size_opt = pointer_offset_size(src.value().type(), ns);
+  // apply 'array-update-with' num_elements times
+  exprt result = src.op();
+
+  std::size_t i = 0;
+  for(const auto &element : value_as_byte_array.operands())
+  {
+    const exprt where = simplify_expr(
+      plus_exprt(src.offset(), from_integer(i, src.offset().type())), ns);
+    ++i;
+
+    // skip elements that wouldn't change (skip over typecasts as we might have
+    // some signed/unsigned char conversions)
+    const exprt &e = skip_typecast(element);
+    if(e.id() == ID_index)
+    {
+      const index_exprt &index_expr = to_index_expr(e);
+      if(index_expr.array() == src.op() && index_expr.index() == where)
+        continue;
+    }
+
+    if(!non_const_update_bound.has_value())
+    {
+      if(result.id() != ID_with)
+      {
+        result = with_exprt(
+          result, where, typecast_exprt::conditional_cast(element, subtype));
+      }
+      else
+      {
+        result.add_to_operands(
+          where, typecast_exprt::conditional_cast(element, subtype));
+      }
+    }
+    else
+    {
+      if_exprt conditional_element(
+        binary_predicate_exprt(
+          from_integer(i, non_const_update_bound->type()),
+          ID_lt,
+          *non_const_update_bound),
+        element,
+        index_exprt(src.op(), where));
+
+      if(result.id() != ID_with)
+      {
+        result = with_exprt(
+          result,
+          where,
+          typecast_exprt::conditional_cast(conditional_element, subtype));
+      }
+      else
+      {
+        result.add_to_operands(
+          where,
+          typecast_exprt::conditional_cast(conditional_element, subtype));
+      }
+    }
+  }
 
-  INVARIANT_WITH_DIAGNOSTICS(
-    element_size_opt.has_value(),
-    "size of type in bytes must be known",
-    irep_pretty_diagnosticst(src));
+  return simplify_expr(result, ns);
+}
 
-  const mp_integer &element_size = *element_size_opt;
+static exprt lower_byte_update_array_vector_non_const(
+  const byte_update_exprt &src,
+  const typet &subtype,
+  const array_exprt &value_as_byte_array,
```

Comment:
PRECONDITION that this really is made of chars?

Reply:
Do you feel this would add much value? I can of course do it, but I think things would either not break or break in an obvious way if that ever were violated.

Reply:
Sure, no big deal